#! /bin/sh -e

# All lines beginning with `# DPATCH:' are a description of the patch.
# DP: PR rtl-optimization/27736
# DP:	* combine.c (replaced_rhs_value): New variable.
# DP:	(combine_instructions): Set it.
# DP:	(distribute_notes): When distributing a note in replaced_rhs_insn,
# DP:	check whether the value was used in replaced_rhs_value.

dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p1 < $0
        #cd ${dir}gcc && autoconf
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p1 < $0
        #rm ${dir}gcc/configure
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0

---

 gcc/combine.c |   10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

Index: gcc-4.1/gcc/combine.c
===================================================================
--- gcc-4.1.orig/gcc/combine.c
+++ gcc-4.1/gcc/combine.c
@@ -128,6 +128,11 @@ static int total_attempts, total_merges,
    so modified, or null if none.  */
 
 static rtx replaced_rhs_insn;
+
+/* When REPLACED_RHS_INSN is nonnull, this is a copy of the new right
+   hand side.  */
+
+static rtx replaced_rhs_value;
 
 /* Vector mapping INSN_UIDs to cuids.
    The cuids are like uids but increase monotonically always.
@@ -883,6 +888,7 @@ combine_instructions (rtx f, unsigned in
 		      rtx orig = SET_SRC (set);
 		      SET_SRC (set) = note;
 		      replaced_rhs_insn = temp;
+		      replaced_rhs_value = copy_rtx (note);
 		      next = try_combine (insn, temp, NULL_RTX,
 					  &new_direct_jump_p);
 		      replaced_rhs_insn = NULL;
@@ -12246,7 +12252,9 @@ distribute_notes (rtx notes, rtx from_in
 	     In both cases, we must search to see if we can find a previous
 	     use of A and put the death note there.  */
 
-	  if (from_insn && from_insn == replaced_rhs_insn)
+	  if (from_insn
+	      && from_insn == replaced_rhs_insn
+	      && !reg_overlap_mentioned_p (XEXP (note, 0), replaced_rhs_value))
 	    tem = from_insn;
 	  else
 	    {
